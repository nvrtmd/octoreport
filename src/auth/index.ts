export * from './login';
export * from './userInfo';
export * from './token';
