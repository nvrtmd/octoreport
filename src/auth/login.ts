import { createOAuthDeviceAuth } from '@octokit/auth-oauth-device';

export async function loginWithGitHubDeviceFlow(
  clientId: string,
  scopes: string[] = ['repo', 'read:user'],
) {
  const auth = createOAuthDeviceAuth({
    clientType: 'oauth-app',
    clientId,
    scopes,
    onVerification: (verification) => {
      console.log('🔐 GitHub authentication is required!');
      console.log(
        `👉 Please visit the following link in your browser: ${verification.verification_uri}`,
      );
      console.log(`✅ Enter the following code in the browser: ${verification.user_code}`);
    },
  });

  const authentication = await auth({ type: 'oauth' });
  return authentication.token;
}
