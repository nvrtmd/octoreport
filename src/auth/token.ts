import keytar from 'keytar';

const { getPassword, setPassword } = keytar;

export async function getGithubToken(githubEmail: string): Promise<string> {
  if (!githubEmail) {
    throw new Error('GitHub email not found. Please log in first.');
  }
  return (await getPassword('octoreport', githubEmail)) ?? '';
}

export async function setGithubToken(githubEmail: string, token: string) {
  if (!githubEmail) {
    throw new Error('GitHub email not found. Please log in first.');
  }
  await setPassword('octoreport', githubEmail, token);
}
