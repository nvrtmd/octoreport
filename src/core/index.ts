export * from './pr/combined';
export * from './pr/userCreated';
export * from './pr/userParticipated';
export * from './pr/filters';
export * from './utils';
