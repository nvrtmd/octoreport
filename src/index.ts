export * from './api';
export * from './core';
export * from './schemas';
export * from './types';
export * from './auth';
