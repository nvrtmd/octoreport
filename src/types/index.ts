export * from './domain';
export * from './input';
