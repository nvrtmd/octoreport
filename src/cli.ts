#!/usr/bin/env node
import inquirer from 'inquirer';

import { loginWithGitHubDeviceFlow } from './auth/login';
import { getToken, setToken } from './auth/token';
import { getUserPRListByCreationAndParticipation } from './core';

const GITHUB_CLIENT_ID = 'Ov23lia7pFpgs8ULT1DL';

if (process.argv[2] === 'login' || !getToken()) {
  const token = await loginWithGitHubDeviceFlow(GITHUB_CLIENT_ID);
  setToken(token);
  console.log('ðŸŽ‰ Authentication successful! You can now use octoreport.');
}

const answers = await inquirer.prompt([
  {
    type: 'input',
    name: 'username',
    message:
      'Optionally, enter the target GitHub username (e.g., octocat) (press Enter to skip and show PRs created by you):',
  },
  {
    type: 'input',
    name: 'repository',
    message: 'Please enter the repository in the format "owner/repo" (e.g., facebook/react):',
  },
  {
    type: 'input',
    name: 'startDate',
    message: 'Please enter the start date for the search period (format: YYYY-MM-DD):',
  },
  {
    type: 'input',
    name: 'endDate',
    message: 'Please enter the end date for the search period (format: YYYY-MM-DD):',
  },
  {
    type: 'input',
    name: 'targetBranch',
    message:
      'Optionally, enter the target branch to filter pull requests (press Enter to skip and show PRs targeting all branches):',
  },
]);

const { userCreatedPRList, userParticipatedPRList } = await getUserPRListByCreationAndParticipation(
  {
    githubToken: getToken() ?? '',
    username: answers.username,
    repository: answers.repository,
    period: {
      startDate: answers.startDate,
      endDate: answers.endDate,
    },
    targetBranch: answers.targetBranch,
  },
);

console.log('User Created PRs: ', userCreatedPRList);
console.log('User Participated PRs: ', userParticipatedPRList);
