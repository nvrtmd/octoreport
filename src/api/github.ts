import { DateTime } from 'luxon';
import { z } from 'zod';

import {
  PRListResponseSchema,
  PRDetailResponseSchema,
  PRListItem,
  PRDetail,
} from '../schemas/github';
import { combinePRData } from '../transformers/github';
import { PRQueryParams, PR, DateRange } from '../types';

function validateApiResponse<T>(data: unknown, schema: z.ZodSchema<T>): T {
  try {
    return schema.parse(data);
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new Error(`API response validation failed: ${error.message}`);
    }
    throw error;
  }
}

export function convertDateToUTCISO(date: string, type: 'start' | 'end' = 'start'): string | null {
  const timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
  const local = DateTime.fromISO(date, { zone: timeZone });
  if (type === 'start') {
    return local.startOf('day').toUTC().toISO();
  }
  return local.endOf('day').toUTC().toISO();
}

export function convertPeriodToUTCISO(period: DateRange): DateRange {
  const startDate = convertDateToUTCISO(period.startDate, 'start') ?? period.startDate;
  const endDate = convertDateToUTCISO(period.endDate, 'end') ?? period.endDate;
  return { startDate, endDate };
}

export async function fetchGitHubUserInfo(
  githubToken: string,
): Promise<{ username: string; email: string }> {
  const response = await fetch('https://api.github.com/graphql', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${githubToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: `
        {
          viewer {
            login
            email
          }
        }
      `,
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const data = await response.json();
  return data.data.viewer;
}

export async function fetchPRListInPeriod(
  options: Pick<PRQueryParams, 'repository' | 'period' | 'githubToken'>,
): Promise<PRListItem[]> {
  let page = 1;
  const prList: PRListItem[] = [];
  const [owner, repo] = options.repository.split('/');

  while (true) {
    const { startDate, endDate } = convertPeriodToUTCISO(options.period);
    const url = `https://api.github.com/search/issues?q=repo:${owner}/${repo}+is:pr+created:${startDate}..${endDate}&sort=created&order=asc&per_page=100&page=${page}`;

    const response = await fetch(url, {
      headers: {
        Authorization: `Bearer ${options.githubToken}`,
        Accept: 'application/vnd.github+json',
      },
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const rawData = await response.json();

    const validatedData = validateApiResponse(rawData, PRListResponseSchema);

    if (validatedData.items.length === 0) {
      break;
    }

    prList.push(...validatedData.items);

    if (!response.headers.get('Link')?.includes('rel="next"')) {
      break;
    }
    page++;
  }

  return prList;
}

export async function fetchPRDetail({
  repository,
  githubToken,
  prNumber,
}: Pick<PRQueryParams, 'repository' | 'githubToken'> & { prNumber: number }): Promise<PRDetail> {
  const [owner, repo] = repository.split('/');
  const query = `query ($owner: String!, $repo: String!, $number: Int!) {
    repository(owner: $owner, name: $repo) {
      pullRequest(number: $number) {
        title
        author { login }
        baseRefName
        assignees(first: 10) { nodes { login } }
        reviews(first: 100) { nodes { author { login }, submittedAt } }
        reviewRequests(first: 10) { nodes { requestedReviewer { ... on User { login } } } }
        comments(first: 10) { nodes { author { login }, createdAt } }
        labels(first: 10) { nodes { name } }
      }
    }
  }`;

  const response = await fetch('https://api.github.com/graphql', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${githubToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables: {
        owner,
        repo,
        number: prNumber,
      },
    }),
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const rawData = await response.json();

  if (rawData.errors) {
    throw new Error(`GraphQL error: ${rawData.errors[0]?.message || 'Unknown error'}`);
  }

  const validatedData = validateApiResponse(rawData, PRDetailResponseSchema);

  return validatedData.data.repository.pullRequest;
}

export async function fetchAllPRListInPeriod(
  options: Pick<PRQueryParams, 'repository' | 'period' | 'githubToken'>,
): Promise<PR[]> {
  const allPRListItems = await fetchPRListInPeriod(options);
  const prList: PR[] = [];
  const { repository, githubToken } = options;

  await Promise.all(
    allPRListItems.map(async (prItem) => {
      try {
        const prDetail = await fetchPRDetail({
          repository,
          githubToken,
          prNumber: prItem.number,
        });
        const prData = combinePRData(prItem, prDetail);
        prList.push(prData);
      } catch (error) {
        console.warn(`⚠️ Failed to process PR #${prItem.number}: ${error}`);
      }
    }),
  );

  prList.sort((a, b) => a.number - b.number);
  return prList;
}
