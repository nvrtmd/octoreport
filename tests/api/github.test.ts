import { describe, expect, it } from 'vitest';

import { fetchPRDetail, fetchPRListInPeriod } from '../../src/api';
import { getGithubToken } from '../../src/auth/token';
import { getUserInfo } from '../../src/auth/userInfo';

const { email: githubEmail } = getUserInfo();
const githubToken = await getGithubToken(githubEmail);

describe('fetchPRListInPeriod', () => {
  it('returns every pull request in the given repository and period', async () => {
    const result = await fetchPRListInPeriod({
      githubToken,
      repository: 'mui/material-ui',
      period: { startDate: '2025-05-01', endDate: '2025-05-31' },
    });

    expect(result).toMatchSnapshot();
  });
});

describe('fetchPRDetail', () => {
  it('returns every pull request detail in the given repository and period', async () => {
    const result = await fetchPRDetail({
      githubToken,
      repository: 'mui/material-ui',
      prNumber: 374,
    });

    expect(result).toMatchSnapshot();
  });
});
