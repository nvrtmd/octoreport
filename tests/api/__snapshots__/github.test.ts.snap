// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchPRDetail > returns every pull request detail in the given repository and period 1`] = `
{
  "assignees": {
    "nodes": [],
  },
  "author": {
    "login": "oliviertassinari",
  },
  "baseRefName": "master",
  "comments": {
    "nodes": [
      {
        "author": {
          "login": "mui-bot",
        },
        "createdAt": "2025-05-02T13:33:29Z",
      },
    ],
  },
  "isDraft": false,
  "labels": {
    "nodes": [
      {
        "name": "website",
      },
      {
        "name": "regression 🐛",
      },
    ],
  },
  "mergeable": "UNKNOWN",
  "merged": true,
  "mergedAt": "2025-05-06T15:03:19Z",
  "reviewDecision": "APPROVED",
  "reviewRequests": {
    "nodes": [],
  },
  "reviews": {
    "nodes": [
      {
        "author": {
          "login": "nadjakovacev",
        },
        "submittedAt": "2025-05-06T12:08:29Z",
      },
    ],
  },
  "state": "MERGED",
  "timelineItems": {
    "nodes": [
      {
        "createdAt": "2025-05-02T22:58:49Z",
        "requestedReviewer": {
          "login": "nadjakovacev",
        },
      },
    ],
  },
  "title": "[website] Update values",
}
`;

exports[`fetchPRListInPeriod > returns every pull request in the given repository and period 1`] = `
[
  {
    "created_at": "2025-04-30T22:17:44Z",
    "html_url": "https://github.com/mui/material-ui/pull/46050",
    "number": 46050,
    "title": "Bump vite [SECURITY]",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-01T00:27:00Z",
    "html_url": "https://github.com/mui/material-ui/pull/46051",
    "number": 46051,
    "title": "[core] Fix internal version duplication",
    "user": {
      "login": "oliviertassinari",
    },
  },
  {
    "created_at": "2025-05-01T03:20:43Z",
    "html_url": "https://github.com/mui/material-ui/pull/46052",
    "number": 46052,
    "title": "Lock file maintenance",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-01T09:36:59Z",
    "html_url": "https://github.com/mui/material-ui/pull/46053",
    "number": 46053,
    "title": "[core] Avoid the need for \`@mui/material/themeCssVarsAugmentation\`",
    "user": {
      "login": "Janpot",
    },
  },
  {
    "created_at": "2025-05-01T14:29:12Z",
    "html_url": "https://github.com/mui/material-ui/pull/46054",
    "number": 46054,
    "title": "[code-infra] Correct some manually curated .d.ts files",
    "user": {
      "login": "Janpot",
    },
  },
  {
    "created_at": "2025-05-01T19:21:03Z",
    "html_url": "https://github.com/mui/material-ui/pull/46055",
    "number": 46055,
    "title": "[website] Adds Rita Iglesias to the team members (About page)",
    "user": {
      "login": "rita-codes",
    },
  },
  {
    "created_at": "2025-05-02T05:23:37Z",
    "html_url": "https://github.com/mui/material-ui/pull/46056",
    "number": 46056,
    "title": "feat(Autocomplete): add \`suppressTextareaWarning\` prop to optionally …",
    "user": {
      "login": "aks30498",
    },
  },
  {
    "created_at": "2025-05-02T06:37:53Z",
    "html_url": "https://github.com/mui/material-ui/pull/46057",
    "number": 46057,
    "title": "[docs][material-ui][Grid] Update grid migration guide",
    "user": {
      "login": "sai6855",
    },
  },
  {
    "created_at": "2025-05-02T07:40:34Z",
    "html_url": "https://github.com/mui/material-ui/pull/46058",
    "number": 46058,
    "title": "[docs][ToggleButtonGroup] Add spacing demo",
    "user": {
      "login": "sai6855",
    },
  },
  {
    "created_at": "2025-05-02T12:53:51Z",
    "html_url": "https://github.com/mui/material-ui/pull/46060",
    "number": 46060,
    "title": "[experiment][core] Upgrade MUI X packages to v8 ",
    "user": {
      "login": "alexfauquette",
    },
  },
  {
    "created_at": "2025-05-02T13:29:40Z",
    "html_url": "https://github.com/mui/material-ui/pull/46061",
    "number": 46061,
    "title": "[website] Update values",
    "user": {
      "login": "oliviertassinari",
    },
  },
  {
    "created_at": "2025-05-03T14:05:12Z",
    "html_url": "https://github.com/mui/material-ui/pull/46064",
    "number": 46064,
    "title": "[docs-infra] Update feedback Node.js to v22",
    "user": {
      "login": "oliviertassinari",
    },
  },
  {
    "created_at": "2025-05-03T15:34:05Z",
    "html_url": "https://github.com/mui/material-ui/pull/46065",
    "number": 46065,
    "title": "[website] Fix CSS alignment",
    "user": {
      "login": "oliviertassinari",
    },
  },
  {
    "created_at": "2025-05-04T03:06:47Z",
    "html_url": "https://github.com/mui/material-ui/pull/46066",
    "number": 46066,
    "title": "Bump @netlify/functions to ^3.1.3",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:08:33Z",
    "html_url": "https://github.com/mui/material-ui/pull/46067",
    "number": 46067,
    "title": "Bump @tailwindcss/postcss to ^4.1.5",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:10:04Z",
    "html_url": "https://github.com/mui/material-ui/pull/46068",
    "number": 46068,
    "title": "Bump @types/react-dom to ^19.1.3",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:11:57Z",
    "html_url": "https://github.com/mui/material-ui/pull/46069",
    "number": 46069,
    "title": "Bump babel to ^7.27.1",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:12:05Z",
    "html_url": "https://github.com/mui/material-ui/pull/46070",
    "number": 46070,
    "title": "Bump github/codeql-action action to v3.28.17",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:13:16Z",
    "html_url": "https://github.com/mui/material-ui/pull/46071",
    "number": 46071,
    "title": "Bump MUI X to 7.29.2",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:14:17Z",
    "html_url": "https://github.com/mui/material-ui/pull/46072",
    "number": 46072,
    "title": "Bump react-virtuoso to ^4.12.7",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:15:35Z",
    "html_url": "https://github.com/mui/material-ui/pull/46073",
    "number": 46073,
    "title": "Bump tailwindcss to ^4.1.5",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:16:41Z",
    "html_url": "https://github.com/mui/material-ui/pull/46074",
    "number": 46074,
    "title": "Bump tsx to ^4.19.4",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:17:46Z",
    "html_url": "https://github.com/mui/material-ui/pull/46075",
    "number": 46075,
    "title": "Bump typescript-eslint to ^8.31.1",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:18:41Z",
    "html_url": "https://github.com/mui/material-ui/pull/46076",
    "number": 46076,
    "title": "Bump mocha to ^11.2.2",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:19:34Z",
    "html_url": "https://github.com/mui/material-ui/pull/46077",
    "number": 46077,
    "title": "Bump express to ^5.1.0",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:20:32Z",
    "html_url": "https://github.com/mui/material-ui/pull/46078",
    "number": 46078,
    "title": "Bump MUI X to 8.2.0 (major)",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:21:22Z",
    "html_url": "https://github.com/mui/material-ui/pull/46079",
    "number": 46079,
    "title": "Bump piscina to ^5.0.0",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:22:10Z",
    "html_url": "https://github.com/mui/material-ui/pull/46080",
    "number": 46080,
    "title": "Bump postcss-combine-media-query to ^2.0.0",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:22:59Z",
    "html_url": "https://github.com/mui/material-ui/pull/46081",
    "number": 46081,
    "title": "Bump react-docgen to ^8.0.0",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-04T03:23:45Z",
    "html_url": "https://github.com/mui/material-ui/pull/46082",
    "number": 46082,
    "title": "Bump react-error-boundary to ^6.0.0",
    "user": {
      "login": "renovate[bot]",
    },
  },
  {
    "created_at": "2025-05-05T04:07:04Z",
    "html_url": "https://github.com/mui/material-ui/pull/46083",
    "number": 46083,
    "title": "[styled-engine] Infer \`ownerState\` from props in \`styled\`",
    "user": {
      "login": "siriwatknp",
    },
  },
  {
    "created_at": "2025-05-05T10:04:20Z",
    "html_url": "https://github.com/mui/material-ui/pull/46084",
    "number": 46084,
    "title": "[performance] Remove unnecessary clsx() usages",
    "user": {
      "login": "sai6855",
    },
  },
  {
    "created_at": "2025-05-05T12:08:49Z",
    "html_url": "https://github.com/mui/material-ui/pull/46085",
    "number": 46085,
    "title": "[website] Add Follower24 as backlink sponsor",
    "user": {
      "login": "rluzists1",
    },
  },
  {
    "created_at": "2025-05-05T13:17:58Z",
    "html_url": "https://github.com/mui/material-ui/pull/46086",
    "number": 46086,
    "title": "[docs-infra] Fix docs API dark mode color",
    "user": {
      "login": "alexfauquette",
    },
  },
  {
    "created_at": "2025-05-05T14:21:13Z",
    "html_url": "https://github.com/mui/material-ui/pull/46087",
    "number": 46087,
    "title": "[release] v7.1.0",
    "user": {
      "login": "siriwatknp",
    },
  },
]
`;
